### PR TITLE
CCS: don't proxy requests for already connected node

### PR DESCRIPTION
**Pull request by javanna**
Cross-cluster search selects a subset of nodes for each remote cluster
and sends requests only to them, which will act as a proxy and properly
redirect such requests to the target nodes that hold the relevant data.
What happens today is that every time we send a request to a remote
cluster, it will be sent to the next node in the proxy list
(in round-robin fashion), regardless of whether the target node is
already amongst the ones that we are connected to. In case for instance
we need to send a shard search request to a data node that's also one of
the selected proxy nodes, we may end up sending the request to it
through one of the other proxy nodes.
This commit optimizes this case to make sure that whenever we are
already connected to a remote node, we will send a direct request rather
than using the next proxy node.
There is a side-effect to this, which is that round-robin will be a bit
unbalanced as the data nodes that are also selected as proxies will
receive more requests.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java [hunk 1]
```diff
@@ -255,40 +253,52 @@ public String executor() {
...
 
     /**
-     * Returns a connection to the remote cluster. This connection might be a proxy connection that redirects internally to the
-     * given node.
+     * Returns a connection to the remote cluster, preferably a direct connection to the provided {@link DiscoveryNode}.
+     * If such node is not connected, the returned connection will be a proxy connection that redirects to it.
      */
     Transport.Connection getConnection(DiscoveryNode remoteClusterNode) {
-        DiscoveryNode discoveryNode = connectedNodes.get();
+        if (connectedNodes.contains(remoteClusterNode)) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java]: should we use `transportService .nodeConnected` instead. I mean if we for instance have the local cluster configured as a remote cluster we get this optimization as well?

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java [hunk 2]
```diff
@@ -255,40 +253,52 @@ public String executor() {
...
-            @Override
-            public DiscoveryNode getNode() {
-                return remoteClusterNode;
-            }
+        return new ProxyConnection(connection, remoteClusterNode);
+    }
 
-            @Override
-            public void sendRequest(long requestId, String action, TransportRequest request, TransportRequestOptions options)
+    static class ProxyConnection implements Transport.Connection {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java]: make the class final?

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java [hunk 3]
```diff
@@ -612,7 +622,7 @@ int getNumNodesConnected() {
         return connectedNodes.size();
     }
 
-    private static class ConnectedNodes implements Supplier<DiscoveryNode> {
+    private static class ConnectedNodes {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java]: also final?

**Review by s1monw — APPROVED** [review]
left some suggestions LGTM

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java]: makes sense, great catch!

**Pull request merged**